internal/handlers/common/projection.go

```diff
@@ -133,102 +171,209 @@ func projectDocument(inclusion bool, doc *types.Document, projection *types.Docu
 			continue
 		}
 
-		switch projectionVal := projectionVal.(type) { // found in the projection
-		case *types.Document: // field: { $elemMatch: { field2: value }}
-			if err := applyComplexProjection(k1, doc, projectionVal); err != nil {
+		switch k1Projection := k1Projection.(type) { // found in the projection
+		case *types.Document: // in projection doc: k1: { k2: value }}, k1Projection == { k2: value }}
+			if err := applyDocProjection(k1, doc, k1Projection); err != nil {
 				return err
 			}
 
-		case float64, int32, int64: // field: number
-			if types.Compare(projectionVal, int32(0)) == types.Equal {
+		case *types.Array: // in projection doc: { k1: [value1, value2... ], k1Projection = [ value1, value2.. ]
+			return NewErrorMsg(ErrNotImplemented, k1+" not supported")
+
+		case float64, // in projection doc: { k1: k1Projection } where k1Projection is a number
+			int32,
+			int64:
+			if types.Compare(k1Projection, int32(0)) == types.Equal {
 				doc.Remove(k1)
 			}
 
-		case bool: // field: bool
-			if !projectionVal {
+		case bool: // in projection doc: { k1: k1Projection }
+			if !k1Projection {
 				doc.Remove(k1)
 			}
 
 		default:
-			return lazyerrors.Errorf("unsupported operation %s %v (%T)", k1, projectionVal, projectionVal)
+			return lazyerrors.Errorf("unsupported operation %s %v (%T)", k1, k1Val, k1Val)
 		}
 	}
 	return nil
 }
 
-func applyComplexProjection(k1 string, doc, projectionVal *types.Document) (err error) {
-	for _, projectionType := range projectionVal.Keys() {
-		supportedProjections := []string{"$elemMatch"}
-		if !slices.Contains(supportedProjections, projectionType) {
-			return fmt.Errorf("projecion %s is not supported", projectionType)
+func applyDocProjection(k1 string, doc *types.Document, k1Projection *types.Document) (err error) {
+	for _, projectionName := range k1Projection.Keys() {
+		if projectionName != "$elemMatch" {
+			panic(projectionName + " not supported!") // checks must be done in projection check func above
 		}
+		var found int
+		conditions := must.NotFail(k1Projection.Get(projectionName)).(*types.Document)
+		found, err = findDocElemMatch(k1, doc, conditions)
+		if err != nil {
+			return
+		}
+		if found < 0 {
+			doc.Remove(k1)
+			return
+		}
+	}
+	return
+}
 
-		// for now it's only $elemMatch further
-		// if the corresponding value is not an array, skip
+func findInArray(k1, k2 string, value any, doc *types.Document, compareRes []types.CompareResult) (found int, err error) {
```

Comment:
Let's not use named parameters as we don't really need them here.

Reply:
done